From 32f985e2a59f09ce34eff1cda00f4d6bee9ae754 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:38:26 -0400
Subject: [virt] kvm/cpuid: Fix sparse warning

Message-id: <1406040016-3289-112-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86190
O-Subject: [RHEL7 PATCH v2 111/221] kvm, cpuid: Fix sparse warning
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Borislav Petkov <bp@suse.de>

We need to copy padding to kernel space first before looking at it.

Reported-by: kbuild test robot <fengguang.wu@intel.com>
Signed-off-by: Borislav Petkov <bp@suse.de>
Signed-off-by: Gleb Natapov <gleb@redhat.com>
(cherry picked from commit 1b2ca42267fe8c4e909312f2b83dfef882cf6234)
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/cpuid.c b/arch/x86/kvm/cpuid.c
index 450f70a..29aa323 100644
--- a/arch/x86/kvm/cpuid.c
+++ b/arch/x86/kvm/cpuid.c
@@ -570,6 +570,7 @@ static bool sanity_check_entries(struct kvm_cpuid_entry2 __user *entries,
      __u32 num_entries, unsigned int ioctl_type)
 {
  int i;
+ __u32 pad[3];
 
  if (ioctl_type != KVM_GET_EMULATED_CPUID)
   return false;
@@ -583,9 +584,10 @@ static bool sanity_check_entries(struct kvm_cpuid_entry2 __user *entries,
   * sheds a tear.
   */
  for (i = 0; i < num_entries; i++) {
-  if (entries[i].padding[0] ||
-      entries[i].padding[1] ||
-      entries[i].padding[2])
+  if (copy_from_user(pad, entries[i].padding, sizeof(pad)))
+   return true;
+
+  if (pad[0] || pad[1] || pad[2])
    return true;
  }
  return false;
-- 
1.7.1